packages/@core.keystone/tasks.js

```diff
@@ -0,0 +1,137 @@
+const Queue = require('bull')
+
+const conf = require('@core/config')
+
+const TASK_TYPE = 'TASK'
+const WORKER_CONCURRENCY = parseInt(conf.WORKER_CONCURRENCY || '2')
+// NOTE: If this is True, all tasks will be executed locally by blocking until the task returns.
+// Tasks will be executed locally instead of being sent to the queue.
+const FAKE_WORKER_MODE = conf.FAKE_WORKER_MODE
+const WORKER_REDIS_URL = conf.WORKER_REDIS_URL
+if (!WORKER_REDIS_URL) throw new Error('No WORKER_REDIS_URL environment')
+const taskQueue = new Queue('tasks', WORKER_REDIS_URL)
+const globalTaskOptions = { attempts: 3, backoff: true }
+const TASKS = new Map()
+
+function createTask (name, fn, opts = {}) {
+    if (typeof fn !== 'function') throw new Error('unsupported fn argument type. Function expected')
+    if (!name) throw new Error('no name')
+
+    if (TASKS.has(name)) throw new Error(`Task with name ${name} is already registered`)
+    TASKS.set(name, fn)
+    return createTaskWrapper(fn, opts)
+}
+
+async function awaitResult (jobId) {
+    const job = await Queue.Job.fromId(taskQueue, jobId)
+    return await job.finished()
+}
+
+function createTaskWrapper (fn, defaultTaskOptions = {}) {
+    async function applyAsync (args, taskOptions) {
+        if (!isSerializable(args)) throw new Error('arguments is not serializable')
+        if (FAKE_WORKER_MODE) {
+            // NOTE: it's just for test purposes
+            // similar to https://docs.celeryproject.org/en/3.1/configuration.html#celery-always-eager
+            // TODO(pahaz): think about test errors!?
+            const result = await executeTask(fn.name, createSerializableCopy([...arguments]))
+            return {
+                getState: async () => { return Promise.resolve('completed') },
+                awaitResult: async () => { return Promise.resolve(result) },
+            }
+        }
+        const job = await taskQueue.add(fn.name, { args: createSerializableCopy([...args]) }, {
+            ...globalTaskOptions,
+            ...defaultTaskOptions,
+            ...taskOptions,
+        })
+        console.log('delayed', fn.name, args, job.id)
```

Comment:
Maybe, it will be better to replace ‘console.log’ to Winston logger, which is configurable depending on environment and have various logs collectors.